src/scanner.cc

```diff
@@ -1,346 +1,228 @@
-//#define DEBUG 1
-#if defined(DEBUG) && DEBUG > 0
- #include <stdio.h> // fprintf
- #define DEBUG_PRINT(fmt, args...) fprintf(stderr, "scanner.cc:%d: " fmt, __LINE__, ##args)
-#else
- #define DEBUG_PRINT(fmt, args...) // dont print
-#endif
+#include <cwctype>
+#include <iostream>
+#include <stdint.h>
+#include <tree_sitter/parser.h>
+#include <vector>
+
+namespace {
+
+using std::vector;
 
 enum TokenType {
-  STRING_FRAGMENT,
-  MULTISTRING_START,
-  MULTISTRING_PART_FIXED,
-  MULTISTRING_IPOL_START,
-  MULTISTRING_END,
+  MULTSTR_START,
+  MULTSTR_END,
+  STR_START,
+  STR_END,
+  INTERPOLATION_START,
+  INTERPOLATION_END,
 };
 
-static void advance(TSLexer *lexer) {
-  lexer->advance(lexer, false);
-}
+struct Scanner {
+  vector<uint8_t> expected_percent_count;
 
-// custom state, shared across calls to scan()
-unsigned int multistring_level = 0;
-int next_token = -1;
-
-// Here we only parse literal fragment inside a string.
-// Delimiter, interpolation and escape sequence are handled by the parser and we simply stop at them.
-//
-// The implementation is inspired by tree-sitter-javascript:
-// https://github.com/tree-sitter/tree-sitter-javascript/blob/fdeb68ac8d2bd5a78b943528bb68ceda3aade2eb/src/scanner.c#L19
-static bool scan_string_fragment(TSLexer *lexer) {
-  DEBUG_PRINT("scan_string_fragment\n");
-  lexer->result_symbol = STRING_FRAGMENT;
-  for (bool has_content = false;; has_content = true) {
-    lexer->mark_end(lexer);
-    switch (lexer->lookahead) {
-      case '"':
-      case '\\':
-        return has_content;
-      case '%':
-        advance(lexer);
-        if (lexer->lookahead == '{') {
-          return has_content;
-        } else if (lexer->lookahead != '"' && lexer->lookahead != '\\') {
-          // Any char following '$' other than '"', '\\' and '{' (which was handled above)
-          // should be consumed as additional string content.
-          // This means `$${` doesn't start an interpolation, but `$$${` does.
-          advance(lexer);
-        }
-        break;
-      // Simply give up on EOF or '\0'.
-      case '\0':
-        return false;
-      default:
-        advance(lexer);
-    }
+  void skip(TSLexer *lexer) { lexer->advance(lexer, true); }
+
+  void advance(TSLexer *lexer) { lexer->advance(lexer, false); }
+
+  int32_t lookahead(TSLexer *lexer) {
+    return lexer->lookahead;
   }
-}
 
-static bool scan_multistring_start(TSLexer *lexer) {
-  DEBUG_PRINT("scan_multistring_start\n");
-  lexer->result_symbol = MULTISTRING_START;
-  // m% was parsed before this
-  // expected: m%" m%%" m%%%" m%%%%" ...
-  // multistring_level = number of %
-  //unsigned int multistring_level = 1;
-  multistring_level = 1;
-  while (lexer->lookahead == '%') {
-    advance(lexer);
-    //lexer->mark_end(lexer); // TODO(milahu) why?
-    multistring_level++;
-    // FIXME skip_unrecognized_character
-    // FIXME lexed_lookahead sym:ERROR, size:3
+  unsigned serialize(char *buffer) {
```

Comment:
rename `buffer` to `state`?

Reply:
These names as taken directly from the [tree-sitter documentation](https://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners). Hence, I'd like to keep them this way.

Reply:
so ... add comments? something like

```cc
// write scanner state to buffer
unsigned serialize(char *buffer) {

// read scanner state from buffer
void deserialize(const char *buffer, unsigned length) {
```